File: packages/typescript/src/catalogs/banxico/tipo-cambio-usd.ts
/**
 * USD/MXN Exchange Rate FIX Catalog
 *
 * Provides access to USD/MXN exchange rate FIX values from Banco de México.
 * The FIX rate is the official exchange rate determined daily by Banxico.
 */

import tipoCambioData from '../../../../shared-data/banxico/tipo_cambio_usd.json';

export interface TipoCambioRecord {
  fecha: string;
  tipo_cambio: number;
  moneda_origen: string;
  moneda_destino: string;
  tipo: string;
  año: number;
  mes: number;
  fuente: string;
}

export class TipoCambioUSDCatalog {
  private static data: TipoCambioRecord[] = tipoCambioData as TipoCambioRecord[];
  private static byFecha: Map<string, TipoCambioRecord> | null = null;
  private static byAnio: Map<number, TipoCambioRecord[]> | null = null;

  private static _loadIndexes(): void {
    if (this.byFecha !== null) return;

    this.byFecha = new Map();
    this.byAnio = new Map();

    for (const record of this.data) {
      this.byFecha.set(record.fecha, record);

      if (!this.byAnio.has(record.año)) {
        this.byAnio.set(record.año, []);
      }
      this.byAnio.get(record.año)!.push(record);
    }

    // Sort by date within each year
    for (const records of this.byAnio.values()) {
      records.sort((a, b) => a.fecha.localeCompare(b.fecha));
    }
  }

  /**
   * Get all exchange rate data
   */
  static getData(): TipoCambioRecord[] {
    return [...this.data];
  }

  /**
   * Get exchange rate for a specific date
   */
  static getPorFecha(fecha: string): TipoCambioRecord | null {
    this._loadIndexes();
    return this.byFecha!.get(fecha) ?? null;
  }

  /**
   * Get all exchange rates for a specific year
   */
  static getPorAnio(anio: number): TipoCambioRecord[] {
    this._loadIndexes();
    return this.byAnio!.get(anio) ?? [];
  }

  /**
   * Get most recent exchange rate
   */
  static getActual(): TipoCambioRecord | null {
    if (this.data.length === 0) return null;
    return this.data.reduce((latest, record) => (record.fecha > latest.fecha ? record : latest));
  }

  /**
   * Get current exchange rate value
   */
  static getValorActual(): number | null {
    const record = this.getActual();
    return record?.tipo_cambio ?? null;
  }

  /**
   * Convert USD to MXN
   */
  static usdAMxn(usd: number, fecha?: string): number | null {
    const record = fecha ? this.getPorFecha(fecha) : this.getActual();
    if (!record) return null;
    return usd * record.tipo_cambio;
  }

  /**
   * Convert MXN to USD
   */
  static mxnAUsd(mxn: number, fecha?: string): number | null {
    const record = fecha ? this.getPorFecha(fecha) : this.getActual();
    if (!record) return null;
    return mxn / record.tipo_cambio;
  }

  /**
   * Calculate percentage variation between two dates
   */
  static calcularVariacion(fechaInicio: string, fechaFin: string): number | null {
    const recordInicio = this.getPorFecha(fechaInicio);
    const recordFin = this.getPorFecha(fechaFin);

    if (!recordInicio || !recordFin) return null;

    return ((recordFin.tipo_cambio - recordInicio.tipo_cambio) / recordInicio.tipo_cambio) * 100;
  }

  /**
   * Calculate annual average exchange rate
   */
  static getPromedioAnual(anio: number): number | null {
    const records = this.getPorAnio(anio);
    if (records.length === 0) return null;

    const sum = records.reduce((acc, r) => acc + r.tipo_cambio, 0);
    return sum / records.length;
  }
}

// Convenience functions
export function getTipoCambioActual(): TipoCambioRecord | null {
  return TipoCambioUSDCatalog.getActual();
}

export function getTipoCambioPorFecha(fecha: string): TipoCambioRecord | null {
  return TipoCambioUSDCatalog.getPorFecha(fecha);
}

export function usdAMxn(usd: number, fecha?: string): number | null {
  return TipoCambioUSDCatalog.usdAMxn(usd, fecha);
}

export function mxnAUsd(mxn: number, fecha?: string): number | null {
  return TipoCambioUSDCatalog.mxnAUsd(mxn, fecha);
}

File: packages/dart/lib/src/catalogs/banxico/tipo_cambio_usd.dart
/// USD/MXN Exchange Rate FIX Catalog
///
/// Catalog of USD/MXN exchange rate FIX values from Banco de México
library;

/// USD/MXN Exchange Rate Catalog
class TipoCambioUsdCatalog {
  static List<Map<String, dynamic>>? _data;

  /// Loads the exchange rate data
  static void _loadData() {
    if (_data != null) return;

    // TODO: Load from packages/shared-data/banxico/tipo_cambio_usd.json
    // In Flutter apps, this would be loaded from assets
    _data = [];
  }

  /// Gets all exchange rate data
  static List<Map<String, dynamic>> getData() {
    _loadData();
    return List.from(_data!);
  }

  /// Gets most recent exchange rate
  static Map<String, dynamic>? getActual() {
    _loadData();
    if (_data!.isEmpty) return null;

    final sorted = List.from(_data!)
      ..sort((a, b) => (b['fecha'] as String).compareTo(a['fecha'] as String));
    return sorted.first;
  }

  /// Gets exchange rate for a specific date
  static Map<String, dynamic>? getPorFecha(String fecha) {
    _loadData();
    return _data!.firstWhere(
      (record) => record['fecha'] == fecha,
      orElse: () => <String, dynamic>{},
    );
  }

  /// Gets current exchange rate value
  static double? getValorActual() {
    final record = getActual();
    return record?['tipo_cambio'] as double?;
  }

  /// Converts USD to MXN
  static double? usdAMxn(double usd, [String? fecha]) {
    final record = fecha != null ? getPorFecha(fecha) : getActual();
    if (record == null || record.isEmpty) return null;

    final rate = record['tipo_cambio'] as double?;
    if (rate == null) return null;

    return usd * rate;
  }

  /// Converts MXN to USD
  static double? mxnAUsd(double mxn, [String? fecha]) {
    final record = fecha != null ? getPorFecha(fecha) : getActual();
    if (record == null || record.isEmpty) return null;

    final rate = record['tipo_cambio'] as double?;
    if (rate == null) return null;

    return mxn / rate;
  }
}
